### ISSUE TITLE
ProtocolException: The server localhost failed to respond with a valid HTTP response

### ISSUE THREAD

**Issue by originalrusyn** [GoogleCloudPlatform/cloud-code-intellij#1072]
Google Cloud SDK 131.0.0
app-engine-java 1.9.44
app-engine-python 1.9.40
beta 2016.01.12
bq 2.0.24
bq-win 2.0.24
bundled-python 2.7.10
core 2016.10.17
core-win 2016.08.05
gcloud
gsutil 4.21
gsutil-win 4.20
powershell 0.1.8
windows-ssh-tools 2016.05.13

jdk 1.8.111

`2016-10-26 09:32:42,558 [1922464006@qtp-461160828-1] DEBUG c.m.c.c.a.AddCreditsController - ggggggggggg aaaa 
validate
validate
����. 26, 2016 9:32:42 AM com.google.appengine.repackaged.org.apache.commons.httpclient.SimpleHttpConnectionManager getConnectionWithTimeout
WARNING: SimpleHttpConnectionManager being used incorrectly.  Be sure that HttpMethod.releaseConnection() is always called and that only one thread and/or method is using this connection manager at a time.
����. 26, 2016 9:32:52 AM com.google.apphosting.utils.jetty.JettyLogger warn
WARNING: /
com.google.appengine.tools.remoteapi.RemoteApiException: remote API call: I/O error
    at com.google.appengine.tools.development.devappserver2.RemoteRpc.makeException(RemoteRpc.java:140)
    at com.google.appengine.tools.development.devappserver2.RemoteRpc.callImpl(RemoteRpc.java:100)
    at com.google.appengine.tools.development.devappserver2.RemoteRpc.call(RemoteRpc.java:48)
    at com.google.appengine.tools.development.devappserver2.RemoteApiDelegate.makeSyncCall(RemoteApiDelegate.java:45)
    at com.google.appengine.tools.development.devappserver2.RemoteApiDelegate$1.call(RemoteApiDelegate.java:56)
    at com.google.appengine.tools.development.devappserver2.RemoteApiDelegate$1.call(RemoteApiDelegate.java:52)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: com.google.appengine.repackaged.org.apache.commons.httpclient.ProtocolException: The server localhost failed to respond with a valid HTTP response
    at com.google.appengine.repackaged.org.apache.commons.httpclient.HttpMethodBase.readStatusLine(Unknown Source)
    at com.google.appengine.repackaged.org.apache.commons.httpclient.HttpMethodBase.readResponse(Unknown Source)
    at com.google.appengine.repackaged.org.apache.commons.httpclient.HttpMethodBase.execute(Unknown Source)
    at com.google.appengine.repackaged.org.apache.commons.httpclient.HttpMethodDirector.executeWithRetry(Unknown Source)
    at com.google.appengine.repackaged.org.apache.commons.httpclient.HttpMethodDirector.executeMethod(Unknown Source)
    at com.google.appengine.repackaged.org.apache.commons.httpclient.HttpClient.executeMethod(Unknown Source)
    at com.google.appengine.repackaged.org.apache.commons.httpclient.HttpClient.executeMethod(Unknown Source)
    at com.google.appengine.tools.development.devappserver2.RemoteRpc.callImpl(RemoteRpc.java:87)
    ... 8 more`

**Comment by patflynn** [comment]
Hi @originalrusyn. Thanks for your report.

I appreciate all the version information you've included of your stack.  Would you be able to provide a bit more information about what you were doing when this error occurred?  Is this on 'Google App Engine Local Development Server' startup?

**Comment by originalrusyn** [comment]
Just run https://github.com/originalrusyn/credits2/tree/googleCloudPlatformIdeaPlugin with Google App Engine Local Development Server plugin for Intellij Idea and open localhost:8080/ several times

**Comment by patflynn** [comment]
@aslo Can you check this out?  I'm wondering if this is an issue in the new dev_appserver.py, or whether this app in particular is just not app engine  compatible.  Running it on the old SDKs dev appserver should be informative.

**Comment by aslo** [comment]
:+1:

**Comment by aslo** [comment]
This does appear to be an issue only with the new dev_appserver implementation. The old SDK dev appserver runs this application without errors. AFAIK, this is a valid app engine application.

My first guess is that something weird is happening with asyncronous objectify requests. I'll investigate further and file a bug with the dev_appserver if necessary.

**Comment by patflynn** [comment]
Thanks Alex.

**Comment by aslo** [comment]
Following up on this issue - it appears that the errors you were seeing regarding `SimpleHttpConnectionManager being used incorrectly`, are indeed related to a bug in the local development server that is used by the google-cloud-intellij plugin. Unfortunately it doesn't look like this issue will be fixed for the time being.

As a workaround, you can use an older version of the App Engine local development server, which is packaged with the [App Engine Java SDK](https://cloud.google.com/appengine/docs/java/download). You can use this development server to manually run and debug your application like so:

```
appengine-java-sdk-1.9.48/bin/dev_appserver.sh /some/path/credits2/target/credits-1.0-SNAPSHOT
```

Closing this issue since it won't be worked on in the near future, and there is a workaround.

**Comment by lborie** [comment]
Hi @aslo,

I have the same blocking issue with my app and Google Cloud Intellij. You have to know that if we want to disable Google Cloud Intellij, and enable the old appengine plugin, it's a bit buggy : it's hard to set the old Java SDK in Intellij.

**Comment by aslo** [comment]
@lborie Thanks for commenting. That's a good point that it's non-trivial to use the old Java SDK from within this intellij plugin. We're going to revisit this issue and see if we can get this issue fixed in the new dev_appserver sooner.

**Comment by lborie** [comment]
Great @aslo! Let me know if I can help you in any way.

**Comment by pmatousek** [comment]
Hi @aslo and @lborie,

I'm experiencing the same issue. It seems that the new dev_appserver's `RemoteRpc` class uses `org.apache.commons.httpclient.HttpClient` with default configuration which internally uses `SimpleHttpConnectionManager` that's not meant to be shared between multiple threads. When the application makes a few concurrent async calls to some App Engine service like putting entities to Datastore or Memcache (using `AsyncDatastoreService`, `AsyncMemcacheService` etc.) these calls are scheduled on different threads in dev_server's `RemoteApiDelegate.makeAsyncCall(...)` but all of them share the same `RemoteRpc` instance. I'm not sure but maybe initializing the `HttpClient` with `MultiThreadedHttpConnectionManager` could help here.

Anyways here's temporary workaround that works for me:

You can replace dev_appserver's `RemoteApiDelegate` with Remote API's `RemoteApiDelegate` and configure it to call local api_server instead.

First make sure that dev_appserver always assigns the same port to the api_server. If you use Gradle with [Google App Engine Gradle plugin](https://github.com/GoogleCloudPlatform/app-gradle-plugin) add the following to your build.gradle (you can choose any port number):

```
appengine {
	run {
		apiPort = 9000
	}
}
```

I'm not sure how it's done in google-cloud-intellij plugin but you can also pass the port directly to dev_appserver.py using --api_port argument.
 
Then add [Remote API](https://mvnrepository.com/artifact/com.google.appengine/appengine-remote-api/latest) library to your project. It's quite big but you don't have to deploy it to App Engine.

Install Remote API before you use any App Engine service. Good place could be your `HttpServlet.init()` method:

```
if (SystemProperty.environment.value().equals(SystemProperty.Environment.Value.Development)) {
	try {
		RemoteApiOptions options = new RemoteApiOptions()
			.server("localhost", 9000)
			.useDevelopmentServerCredential();

		new RemoteApiInstaller().installOnAllThreads(options);
	} catch (IOException e) {
		logger.log(Level.SEVERE, "failed to install remote API", e);
	}
}
```

Make sure that dev_appserver assigns proper app id to your application by specifying `<application>my-app-id</application>` element in your appengine-web.xml or passing it to dev_appserver.py using -A argument. Otherwise it sets dev~None as app id which makes the Remote API installation to fail.

That's all. Hope this helps.

**Comment by eamonnmcmanus** [comment]
@pmatousek can you explain why you say that the different threads are using the same `RemoteRpc` instance? I can't see how that would be from looking at the code but perhaps I am missing something.

**Comment by pmatousek** [comment]
Hi @eamonnmcmanus,

I'll try but I have only devserver binary from gcloud so it may not be very accurate. I use Google Cloud SDK 140.0.0 and App Engine Java libraries 1.9.48.

Here's a sample piece of code that causes the exception almost 100%:

```
AsyncDatastoreService ads = DatastoreServiceFactory.getAsyncDatastoreService();
Future<?> f1 = ads.put(new Entity("A", 1));
Future<?> f2 = ads.put(new Entity("B", 1));

f1.get();
f2.get(); // exception is thrown
```

I think that internally happens the following:

There're 3 key classes. `DevAppServer2Delegate`, `RemoteApiDelegate` and `RemoteRpc` from `com.google.appengine.tools.development.devappserver2` package.

After startup devserver installs `DevAppServer2Delegate` as it's `ApiProxy.delegate`.

This `DevAppServer2Delegate` has a `ThreadLocal` type field that holds instances of `RemoteApiDelegate`.

`RemoteApiDelegate` internally uses `ExecutorService` with fixed thread pool to schedule calls to it's single `RemoteRpc` instance when you call `RemoteApiDelegate.makeAsyncCall(...)`.

The `RemoteRpc` class internally uses `HttpClient` with default configuration that has `SimpleHttpConnectionManager` which is not supposed to be used by multiple threads. If used this way there should be probably separate `RemoteRpc` instance for every thread spawned by `ExecutorService` in `RemoteApiDelegate` or `HttpClient` should be configured differently.

Then the call `f1/f2.get()` gets to the point where it calls

```
1) DatastoreApiHelper.makeAsyncCall(...) // line 107
2) ApiProxy.makeAsyncCall(...) // line 201
3) DevAppServer2Delegate.makeAsyncCall(...) // line 73 - uses the same RemoteApiDelegate from ThreadLocal field
4) RemoteApiDelegate.makeAsyncCall(...) // line 52 - schedules call to the same RemoteRpc instance from different threads
```

Since both `f1/f2.get()` are called from the same thread both of them at point 3) get the same `RemoteApiDelegate` from `DevAppServer2Delegate.threadLocalRemoteApiDelegate`.

As I wrote above although this `RemoteApiDelegate` schedules the calls to different threads using it's `ExecutorService` it doesn't matter since these threads share the same `RemoteRpc` instance.
You can even check that the `RemoteRpc` instance is reused as it's `rpcCount` field gets incremented after each call and during `f2.get()` the `rpcCount` value is already 1.

I'm sorry if I made this too complicated to follow. Please let me know if I should explain some parts more clearly.

**Comment by eamonnmcmanus** [comment]
Oh, that's very helpful! The fundamental issue is that the code assumes synchronous calls. Asynchronous calls can lead to a request being sent from a thread before a previous request from the thread has terminated. The async aspect wasn't obvious from the earlier stack trace, which may represent a synchronous call happening on a thread that already has an unfinished asynchonous request outstanding.

**Comment by pmatousek** [comment]
@eamonnmcmanus You're welcome. Yes you're right that the async aspect of this is somewhat hidden in the callstack - it wasn't obvious to me until I did some deeper debugging. Unfortunately it's sometimes not easy to avoid asynchronous calls. Maybe it might interest you that I ran into a similar problem recently - I described it in more detail in a [Google cloud platform issue tracker](https://code.google.com/p/google-cloud-platform/issues/detail?id=207).
